### PR TITLE
Issue 44632: Publish study wizard support for selecting user-defined Queries

### PR DESCRIPTION
**Pull request by cnathe**
#### Rationale
https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=44632

Study publish wizard exports & imports every archivable object we have... except queries. That means, for example, that publishing the LabKey demo study results in broken R charts (because they depend on queries). As far as we can tell, the intent was to show queries on their own panel (allowing selecting subsets to publish), but it was never actually wired up. This PR does just that.

#### Related Pull Requests
* https://github.com/LabKey/platform/pull/3034
* https://github.com/LabKey/testAutomation/pull/1004

#### Changes
* update wizard to add a new panel that shows a grid of the user-defined queries for all schemas in the given container
* GetSchemaQueryTreeAction updates to get all user defined queries for a container
* pipe through query key selections to CreateChildStudyPipelineJob and FolderExportContext to be used in QueryWriter
* handle query key selections in study.StudySnapshot definitions
* update the set of Grid Views for selection based on the selected query keys (in addition to the already supported list and dataset selection filtering)

### CODE REVIEW

File: query/src/org/labkey/query/QueryWriter.java [hunk 1]
```diff
@@ -68,6 +70,13 @@ public void write(Container object, ImportContext<FolderDocument.Folder> ctx, Vi
         List<QueryDefinition> queries = new ArrayList<>(QueryServiceImpl.get().getQueryDefsAndMetadataOverrides(ctx.getUser(), c));
         FileNameUniquifier fileNameUniquifier = new FileNameUniquifier();
 
+        if (ctx.getClass().equals(FolderExportContext.class))
+        {
+            Set<String> queryKeysToExport = ((FolderExportContext)ctx).getQueryKeys();
+            if (queryKeysToExport != null)
+                queries.removeIf(queryDef -> !queryKeysToExport.contains(queryDef.getQueryKey()));
+        }
```
Comment (labkey-tchad) [inline query/src/org/labkey/query/QueryWriter.java]: Why exclude `FolderExportContext` subclasses? Is there a chance they would want this behavior?

```suggestion
        if (ctx instanceof FolderExportContext feCtx)
        {
            Set<String> queryKeysToExport = feCtx.getQueryKeys();
            if (queryKeysToExport != null)
                queries.removeIf(queryDef -> !queryKeysToExport.contains(queryDef.getQueryKey()));
        }

```

File: study/test/src/org/labkey/test/tests/study/StudyPublishQueriesTest.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        // need at least one visit for the publish study wizard
+        _studyHelper.goToManageVisits().goToImportVisitMap();
+        setFormElement(Locator.name("content"),
+        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
+                "<visitMap xmlns=\"http://labkey.org/study/xml\">\n" +
+                "  <visit label=\"Test Visit1\" typeCode=\"X\" sequenceNum=\"1.0\" maxSequenceNum=\"1.0\"/>\n" +
+                "  <visit label=\"Test Visit2\" typeCode=\"X\" sequenceNum=\"2.0\" maxSequenceNum=\"2.0\"/>\n" +
+                "  <visit label=\"Test Visit3\" typeCode=\"X\" sequenceNum=\"3.0\" maxSequenceNum=\"3.0\"/>\n" +
+                "</visitMap>"
+        );
```
Comment (labkey-tchad) [inline study/test/src/org/labkey/test/tests/study/StudyPublishQueriesTest.java]: Java 16 text blocks!!
```suggestion
        setFormElement(Locator.name("content"),
        """
        <?xml version=\"1.0\" encoding=\"UTF-8\"?>
          <visitMap xmlns=\"http://labkey.org/study/xml\">
            <visit label=\"Test Visit1\" typeCode=\"X\" sequenceNum=\"1.0\" maxSequenceNum=\"1.0\"/>
            <visit label=\"Test Visit2\" typeCode=\"X\" sequenceNum=\"2.0\" maxSequenceNum=\"2.0\"/>
            <visit label=\"Test Visit3\" typeCode=\"X\" sequenceNum=\"3.0\" maxSequenceNum=\"3.0\"/>
          </visitMap>
        """
        );
```

File: study/test/src/org/labkey/test/tests/study/StudyPublishQueriesTest.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        // create user defined queries of the same name in different schemas
+        for (String schemaName : SCHEMA_NAMES)
+            createQuery(getProjectName(), USER_DEFINED_QUERY, schemaName, "SELECT 'Test from " + schemaName + " schema'", null, false);
+    }
+
+    @Override // TODO delete me
+    protected void doCleanup(boolean afterTest)
+    {
+        super.doCleanup(afterTest);
+    }
```
Comment (labkey-tchad) [inline study/test/src/org/labkey/test/tests/study/StudyPublishQueriesTest.java]: Delete me!

**Review by labkey-tchad — APPROVED** [review]

**Review by labkey-danield — APPROVED** [review]

**Review by labkey-adam — APPROVED** [review]

File: study/test/src/org/labkey/test/tests/study/StudyPublishQueriesTest.java (hunk 2 above)
Comment (cnathe) [reply study/test/src/org/labkey/test/tests/study/StudyPublishQueriesTest.java]: Cool! Done.

File: study/test/src/org/labkey/test/tests/study/StudyPublishQueriesTest.java (hunk 3 above)
Comment (cnathe) [reply study/test/src/org/labkey/test/tests/study/StudyPublishQueriesTest.java]: Thanks, done.

File: query/src/org/labkey/query/QueryWriter.java (hunk 1 above)
Comment (cnathe) [reply query/src/org/labkey/query/QueryWriter.java]: Good point. I was copying over logic from the ReportWriter. I've updated both to your suggestion. Thanks.

**Review by labkey-tchad — APPROVED** [review]

**Pull request merged**